### DISCUSSION TITLE
GraphQL query and C#

### DISCUSSION THREAD

**Discussion by BumpaRoy** [Azure/data-api-builder#1732] [Q&A] [answered]
Not sure if you can answer this.  But running the query on Postman returns data, but when I run it in C# it doesn't.  Any ideas?

`        public static async Task RunAsync()

        {
            // Update port # in the following line.
            client.BaseAddress = new Uri(ServerUrl);
            client.DefaultRequestHeaders.Accept.Clear();
            client.DefaultRequestHeaders.Accept.Add(
                new MediaTypeWithQualityHeaderValue("application/json"));

            // GET: Get data from Server
            var productRequest = new GraphQLRequest
            { Query = @"{
                products
                {
                    items {
                        fpartno
                        frev
                        fdescript
                    }
                }
            }" };
            //var productRequest = new GraphQLRequest
            //{ Query = @"{ products(filter:{fpartno: {contains: ""WF""}}) { items { fpartno frev fdescript } }" };

            var graphQLClient = new GraphQLHttpClient(ServerUrl, new NewtonsoftJsonSerializer());
            var graphQLResponse = await graphQLClient.SendQueryAsync<Rootobject>(productRequest);

            Debug.Print("");

        }`

**Comment by seantleonard** [comment]
Hard to say with the information provided. Can you provide more Info about
- GraphQL client are you using
- Any other response properties like errors or HTTP response codes for the query that fails
- Has stepping through in the debugger provided any helpful insights
- What is `Rootobject` in your code

**Reply by BumpaRoy** [reply]
Errors appear to be Null (Pic02), Data is Null (Pic03) and the client options is shown in Pic01.  Rootobject is the return structure I am expecting to be returned.  Used the Paste JSON as Classes in VS2022.

Here is the whole class:
`using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;
using GraphQL;
using GraphQL.Client;
using System.Runtime.CompilerServices;
using GraphQL.Client.Http;
using GraphQL.Client.Serializer.Newtonsoft;

namespace GraphQLMain
{

    public class Rootobject
    {
        public Data data { get; set; }
    }

    public class Data
    {
        public List<Products> products { get; set; }
    }

    public class Products
    {
        public Item[] items { get; set; }
    }

    public class Item
    {
        public string fpartno { get; set; }
        public string frev { get; set; }
        public string fdescript { get; set; }
    }
    public class GraphQLClass
    {

        public static async Task RunAsync()

        {
            var graphQLClient = new GraphQLHttpClient("https://localhost:5001/graphql", new NewtonsoftJsonSerializer());

            var productRequest = new GraphQLRequest
            { Query = @"{
                products
                {
                    items {
                        fpartno
                        frev
                        fdescript
                    }
                }
            }" };
            //var productRequest = new GraphQLRequest
            //{ Query = @"{ products(filter:{fpartno: {contains: ""WF""}}) { items { fpartno frev fdescript } }" };

            var graphQLResponse = await graphQLClient.SendQueryAsync<Rootobject>(productRequest);

            Debug.Print("");

        }
    }
}`
![graphQL02](https://github.com/Azure/data-api-builder/assets/137801704/1b630180-729c-4306-a845-abd9ac8c9828)
![graphQL01](https://github.com/Azure/data-api-builder/assets/137801704/242fcde0-cbda-47f0-82ff-1009b2db0ee5)
![graphQL03](https://github.com/Azure/data-api-builder/assets/137801704/53a4c2dc-27bc-4fed-930a-8ae36918de6a)

**Reply by BumpaRoy** [reply]
Here is the Postman query:
![graphQL04](https://github.com/Azure/data-api-builder/assets/137801704/9716870c-cc64-49a6-bd44-0a2e56031ffe)

**Comment by Aniruddh25** [comment]
hi @BumpaRoy,

Thank you for the detailed information. Seems like an issue with deserialization from the returned response. Is it possible to check what the raw JSON content response is as returned from the server using your GraphQL.Client? Like @seantleonard asked earlier, do you have any links to which library you are getting `GraphQL.Client` from?

Another option is to try executing your GraphQL request as we do in our tests as follows: 
https://github.com/Azure/data-api-builder/blob/main/src/Service.Tests/GraphQLRequestExecutor.cs
Once you get the raw JSON content, you can deserialize into your `RootObject` as a post processing step.

Thanks!

**Comment by BumpaRoy** [comment]
I am sorry, but I don't know how to capture the raw json content.  The value of the GraphQLClient.  Not sure if the screen print helps.  As far as the client version it is the latest from gitHub.
![2023-09-25_10-31-25](https://github.com/Azure/data-api-builder/assets/137801704/0fd8d975-d8c7-459f-9d35-4151f4f045b3)

**Reply by BumpaRoy** [reply]
Here is the log entries from the DAB when I run the query.  It is spot on from the way it is generating the select.

SELECT TOP 100 [table0].[fpartno] AS [fpartno], [table0].[frev] AS [frev] FROM [dbo].[inmastx] AS [table0] WHERE 1 = 1 ORDER BY [table0].[identity_column] ASC FOR JSON PATH, INCLUDE_NULL_VALUES
dbug: Azure.DataApiBuilder.Core.AuthenticationHelpers.ClientRoleHeaderAuthenticationMiddleware[0]
      b4207740-ba07-41e3-9b14-ee8afce9ecb5: The request will be executed in the context of Anonymous role
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'Hot Chocolate GraphQL Pipeline'
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/1.1 POST https://localhost:5001/graphql/ application/json;+charset=utf-8 288 - 200 - application/json;+charset=utf-8 8.6947ms

**Reply by BumpaRoy** [reply]
Any more thoughts on this.  The RootObject is being returned as Null with no errors.  I can make it return an error with a bad query syntax, but this is being returned with Null Data and Null for Errors as well.  It is being sent to DAB as the log shows that above.  I am at a dead end.  Any ideas would be appreciated.

**Reply by seantleonard** [reply, marked as answer]
This seems to be an issue with how `Newtonsoft.Json` deserializes the response into your `RootObject` class, and not something that you need to configure within DAB. 
Troubleshooting suggestion 1:

I think that your `Data` class needs to change because the GraphQL response property `products` is not a list, it's the name of the GraphQL query you defined in your request:
```csharp
public class RootObject
{
    public Products products { get; set; }
}
```

I'm skipping the class `Data` because that class doesn't look to be used in the GraphQL.net library you use for deserializing a GraphQL response into a particular object.

More example code that your GraphQL client provides: https://github.com/graphql-dotnet/graphql-client#operationname-and-variables-request

Troubleshooting suggestion 2:

You could also try debugging with modifying your request save the response as `object` to then use the debugger to check the output to confirm whether your class structure needs adjustment:
```csharp
        var graphQLResponse = await graphQLClient.SendQueryAsync<object>(productRequest);
```

**Reply by BumpaRoy** [reply]
That worked!!  Thank you so much!!!  I used the Paste JSON as classes from the Postman results and it created the Class structure I was using.  I just assumed that to be correct.  The suggestion of returning it to the object is also a great tip.  Thanks Again!  Now onto the next hurdle ..
